### PR TITLE
Implements Perlin noise based map generation and jungle generation

### PR DESCRIPTION
**Pull request by Sovexe**
## About the PR 
Ports https://github.com/tgstation/tgstation/pull/51082 with author's permission.

> the generation basically works like this: We have 3 layers of perlin noise (a form of noise that seamlessly transitions between 0 to 1) on top of each other, the first is height, if its above 0.85 it'll be rock, if it's lower, it'll use heat and humidity, the other 2 layers, to pick a biome from a 3x3 selection.

examples:
https://streamable.com/o895v9
image

old examples:
https://streamable.com/ypluyn
https://streamable.com/z51b3h

This could seriously benefit from additional objects and turfs. I picked from what we have best I could.

## Why's this needed? 
* It's cool

### CODE REVIEW

File: _std/map.dm [hunk 1]
```diff
@@ -6,3 +6,18 @@
...
+
+#define BIOME_LOW_HEAT "low_heat"
+#define BIOME_LOWMEDIUM_HEAT "lowmedium_heat"
+#define BIOME_HIGHMEDIUM_HEAT "highmedium_heat"
+#define BIOME_HIGH_HEAT "high_heat"
+
+#define BIOME_LOW_HUMIDITY "low_humidity"
+#define BIOME_LOWMEDIUM_HUMIDITY "lowmedium_humidity"
+#define BIOME_HIGHMEDIUM_HUMIDITY "highmedium_humidity"
+#define BIOME_HIGH_HUMIDITY "high_humidity"
```
Comment (Adharainspace) [inline _std/map.dm:23]: these defines look good to me :]

File: code/datums/mapgen/_MapGenerator.dm [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+///Initialize all biomes, assoc as type || instance
+proc/initialize_biomes()
+	for(var/biome_path in concrete_typesof(/datum/biome))
+		var/datum/biome/biome_instance = new biome_path()
+		biomes[biome_path] += biome_instance
+
+///This type is responsible for any map generation behavior that is done in areas, override this to allow for area-specific map generation. This generation is ran by areas in initialize.
+/datum/map_generator
+
+///This proc will be ran by areas on Initialize, and provides the areas turfs as argument to allow for generation.
```
Comment (ZeWaka) [inline code/datums/mapgen/_MapGenerator.dm:13]: New

File: code/datums/mapgen/biomes/_biome.dm [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	///Type of turf this biome creates
+	var/turf_type
+	///Chance of having a structure from the flora types list spawn
+	var/flora_density = 0
+	///Chance of having a mob from the fauna types list spawn
+	var/fauna_density = 0
+	///list of type paths of objects that can be spawned when the turf spawns flora
+	var/list/flora_types = list(/obj/tree1)
+	///list of type paths of mobs that can be spawned when the turf spawns fauna
+	var/list/fauna_types = list()
```
Comment (ZeWaka) [inline code/datums/mapgen/biomes/_biome.dm:12]: would love for these to be weighted assoc. lists

File: code/datums/mapgen/_MapGenerator.dm [hunk 4]
```diff
@@ -0,0 +1,45 @@
+///All possible biomes in assoc list as type || instance
+var/list/biomes = list()
```
Comment (ZeWaka) [inline code/datums/mapgen/_MapGenerator.dm:2]:
```suggestion
var/list/datum/biome/biomes = list()
```

File: code/datums/mapgen/_MapGenerator.dm [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+///Initialize all biomes, assoc as type || instance
+proc/initialize_biomes()
+	for(var/biome_path in concrete_typesof(/datum/biome))
+		var/datum/biome/biome_instance = new biome_path()
+		biomes[biome_path] += biome_instance
+
+///This type is responsible for any map generation behavior that is done in areas, override this to allow for area-specific map generation. This generation is ran by areas in initialize.
```
Comment (ZeWaka) [inline code/datums/mapgen/_MapGenerator.dm:10]:
```suggestion
///This type is responsible for any map generation behavior that is done in areas, override this to allow for area-specific map generation. This generation is ran by areas in New.
```

**Review by ZeWaka — APPROVED** [review]

File: code/datums/mapgen/_MapGenerator.dm (hunk 5 above)
Comment (Sovexe) [reply code/datums/mapgen/_MapGenerator.dm:10]: Oh I need to fix that comment. I ended up not using New as the biomes weren't initialized in time for the area New proc to run

**Comment by Sovexe** [comment]
made the water nicer 
image

File: code/datums/mapgen/biomes/_biome.dm (hunk 3 above)
Comment (Sovexe) [reply code/datums/mapgen/biomes/_biome.dm]: added

**Comment by Sovexe** [comment]
Merging this as it is for now since it's usable for gimmicks and early testing.

**Pull request merged**